### PR TITLE
Support Trait

### PR DESCRIPTION
**Pull request by yongpingchen**
Introduction refers: https://github.com/KiiPlatform/thing-if-JSSDK/pull/155

#### Changes
- Remove schema and schema version
  - remove Schema and SchemaBuilder class
  - remove schema and schemaVersion parameters in ThingIFAPI#postNewCommand,  ThingIFAPI#postNewTrigger and ThingIFAPI#patchTrigger
  - remove schema and schemaVersion from ThingIFAPIBuilder
  - remove schema and schemaVersion from Command, CommandForm and TriggeredCommandForm. 
  - remove schema from GsonRepository class
- support post trait formatted command(use `ThingIFAPI#postNewCommand(CommandForm form)`)
  - add `TraitActions` and `TraitEnabledCommandActions` class to represent trait format actions in command. 
  - add `TraitActionResults` and `TraitEnabledActionResults` to represent trait format action results in command.
  - overload constructor of `Command` with `TraitEnabledCommandActoins` instance
  - overload constructor of `CommandForm` with `TraitEnabledCommandActoins` instance
- support trait formatted trigger
   Use the following methods for trait formatted trigger:  `ThingIFAPI#postNewTrigger(TriggerCommandForm form, Predicate predicate, TriggerOptions options)`, `ThingIFAPI#patchTrigger(String triggerID, TriggerCommandForm form, Predicate predicate, TriggerOptions options)`
  - overload TriggerCommandForm.Builder#newBuilder method with `TraitEnabledCommandActoins` instance
  - Add trait alias in clauses: Equals, Range
- Overload getTargetState with alias and firmwareVersion parameters to get states with specified alias/firmwareVersion. 
- Add APIs to query history states of thing
  - add `ThingIFAPI#queryHistoryStates` to query history states.
  - add `ThingIFAPI#aggregateHistoryStates` to aggregate history states. 
  - add `HistoryStateQuery` to represent query history states for the above 2 methods.
  - add TimeRange clause for whitInRange query 
  - add `HistoryStateResponse` as class of return value of `ThingIFAPI#queryHistoryState` method
  - add `AggregatedHistoryStatesResponse` as class of return value of `ThingIFAPI#aggregateHistoryStates`, and `AggregationResult` class.
  - Add Aggregation and AggregationType for aggregations optional parameter when query history state. 
- Add ThingIFAPI#updateFirmwareVersion
- Add ThingIFAPI#updateThingType for updating thingType to make the thing starting to support trait.

### CODE REVIEW

File: thingif/src/main/java/com/kii/thingif/command/CommandForm.java [hunk 1]
```diff
@@ -22,23 +20,18 @@
...
+ * <li>meta data of a command</li>
  * </ul>
  */
 public final class CommandForm implements Parcelable {
 
-    @SerializedName("schema")
-    private final @NonNull String schemaName;
-    private final int schemaVersion;
-    private final @NonNull List<Action> actions;
+    private final @NonNull List actions;
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/command/CommandForm.java]: I think that generic without specifying type is not good manner.

I think that introducing interface and implementing the interface at `Action` and `TraitActions` is resolve the issue. What do you think?

File: thingif/src/main/java/com/kii/thingif/ThingIFAPI.java [hunk 2]
```diff
@@ -815,16 +787,12 @@ public Command getCommand(
     @NonNull
     @WorkerThread
     public Trigger postNewTrigger(
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/ThingIFAPI.java]: Non `TriggeredCommandForm` version `postNewTrigger` and `patchTrigger` were removed in thing-if iOSSDK.
If trait adaptable thing-if AndroidSDK becomes version 1.0.0, Could we remove such duplicated methods like thing-if iOSSDK done?

File: thingif/src/main/java/com/kii/thingif/query/Aggregation.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
+package com.kii.thingif.query;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+
+/**
+ * Represent aggregation
+ */
+public class Aggregation {
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/query/Aggregation.java]: @yongpingchen 

Don't `Aggregation` need to implement `Parcelable`?

File: thingif/src/main/java/com/kii/thingif/query/Aggregation.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+     * @param type Type of aggregation
+     * @param responseField
+     * @param field
+     * @param fieldType
+     */
+    public Aggregation(
+            AggregationType type,
+            String responseField,
+            String field,
+            String fieldType
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/query/Aggregation.java]: @yongpingchen 

Could you add Nullable/NonNull annotation?

File: thingif/src/main/java/com/kii/thingif/ThingIFAPI.java (hunk 2 above)
Comment (yongpingchen) [inline thingif/src/main/java/com/kii/thingif/ThingIFAPI.java]: I agree to remove this kinds of methods.

File: thingif/src/main/java/com/kii/thingif/query/Aggregation.java (hunk 3 above)
Comment (yongpingchen) [inline thingif/src/main/java/com/kii/thingif/query/Aggregation.java]: Should be

**Comment by yongpingchen** [comment]
New design applied on branch `support_trait_design`. Will not work on this PR any more.

**Pull request closed without merging**